### PR TITLE
Create 104-dht-deep-dive.md

### PR DESCRIPTION
**Pull request by momack2**
What may be our largest blog post yet diving deep on how the IPFS DHT was upgraded in H1 2020.

todo:
- add links!
- add images
- audit for formatting issues
- insert commas
- TBD: talk about how we've actually now met our IPFS H1 content routing performance goal of <5second 95th percentile content routing! 🎉

### CODE REVIEW

File: content/post/104-dht-deep-dive.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+
+## Parting Thoughts
+
+Phew! If you have made it all the way through this blog post (or even just skimmed most of it), I commend you! Before sending you back to your exciting lives — just a few brief comments about IPFS v0.5.0 and the releases to come:
+
+IPFS v0.5.0 included a lot of DHT changes and improvements. Something to watch out for is that new nodes are currently participating in the same DHT as older go-ipfs v0.4.23 and earlier nodes. While the DHT code for v0.5.0 is much improved over previous versions, a single node in a big network can only help so much. This means you should see improved performance today when you run v0.5.0, but as more of the network upgrades to v0.5.0 and beyond we will continue to see lookup times improve. So tell your friends it’s time to upgrade! 😁
+
+There are many more exciting improvements to come - so if you are interested in contributing or just tracking our improvements, follow our repositories on [GitHub](https://github.com/ipfs/ipfs) and come chat with us on [IRC](https://github.com/ipfs/team-mgmt#chat).
+
```
Comment (momack2) [inline content/post/104-dht-deep-dive.md:150]:
```suggestion
### Learn more

* IPFS 0.5.0 Announcement: https://blog.ipfs.io/2020-04-28-go-ipfs-0-5-0/
* Release Highlights:  https://www.youtube.com/watch?v=G8FvB_0HlCE
* TestGround: https://blog.ipfs.io/2020-05-06-launching-testground/
```

**Comment by momack2** [comment]
Circle CI is unhappy - something about un-marshaling yaml (potentially related to combining bullets and numbered lists?) - will debug later.

File: content/post/104-dht-deep-dive.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
+---
+date: 2020-07-20
+url: /2020-07-20-dht-deep-dive/
+title: IPFS 0.5 Content Routing Improvements: Deep Dive
```
Comment (olizilla) [inline content/post/104-dht-deep-dive.md:4]: @momack2 The error is complaining about this line... I suspect the colon in `Improvements: Deep` is being interpretreted as attepting to write a nested object.

```
Error: Error building site: "/home/circleci/project/content/post/104-dht-deep-dive.md:4:1": failed to unmarshal YAML: yaml: line 3: mapping values are not allowed in this context
 ```

Comment (olizilla) [reply content/post/104-dht-deep-dive.md:4]: I think wrapping the the title in quotes `"` should fix it

```suggestion
title: "IPFS 0.5 Content Routing Improvements: Deep Dive"
```

File: content/post/104-dht-deep-dive.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+1. When we connect to a peer, check if it qualifies to be added to our routing table
+    - Ensure the peer is a DHT server that is advertising the DHT protocol ID (in IPFS `/ipfs/kad/1.0.0` for the public/WAN DHT and `/ipfs/lan/kad/1.0.0` for the LAN DHT)
+    - Ensure the peer has IP addresses that match the ranges we expect (e.g. members of the public DHT having at least one public range IP address as opposed to only addresses like `192.168.X.Y`)
+2. If it qualifies, then determine how close the new peer is in the Kademlia address space to us to figure out which “bucket” it should go into (i.e. if the peer is between 2^7 and 2^8 away from us and the address space is of size 2^256 then the peer goes into bucket 256-8)
+3. Attempt to put the peer in the bucket
+    - If the bucket is not full (i.e., has less than 20 peers in it) then add the peer
+    - If the bucket is full then determine if there are any peers that have not been “recently useful” to us, and then drop one of them and replace it with the new peer (“recently” and “useful” are defined below). Otherwise, don’t add the peer to the bucket
```
Comment (aschmahmann) [inline content/post/104-dht-deep-dive.md:81]: We've somewhat tossed/modified this concept since the release of v0.5. @petar what do you think about changing this to:

```suggestion
    - If the bucket is full then determine if there are any peers that are "replaceable" (defined below), and then drop one of them and replace it with the new peer. Otherwise, don’t add the peer to the bucket
```

File: content/post/104-dht-deep-dive.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+5. After the query is done take the `K` closest peers that we have not already failed at querying (i.e. we have heard from them or they are still in our queue) and return them
+    - Note: for some API compatibility reasons go-ipfs also ensures that we have actually sent queries to all of the top `K` peers
+
+In the routing table section, we mentioned that we evict peers that have not been “recently useful” to us if we find a new peer that could take their spot in the bucket.
+We define “recently” as the time period during which they are probabilistically expected to have been utilized in a refresh. That value is `Log(1/K) * Log(1 - Alpha/K) * refreshPeriod`, where Alpha is the number of peers dialed that can be simultaneously queried. Additionally, we define “useful” as responding within 2x the time it takes any other peer from our routing table to respond to us (this biases against peers that are slow, overloaded, unreliable, or have bad network connectivity to us). This definition of useful will likely change as we gather more information about the dynamics of the network and investigate relevant threat models.
```
Comment (aschmahmann) [inline content/post/104-dht-deep-dive.md:105]: @petar If we change the corresponding section above then maybe something like the below (especially given that we have a different model, i.e. replaceable peers are only ones that we found during bootstrapping, scheduled for the next release)

```suggestion
In the routing table section, we mentioned that we evict peers that are "replaceable" if we find a new peer that could take their spot in the bucket.
In v0.5.0 we define a peer as "replaceable" if they have not been "useful" to us within the time period during which they are probabilistically expected to have been utilized in a refresh. That value is `Log(1/K) * Log(1 - Alpha/K) * refreshPeriod`, where Alpha is the number of peers dialed that can be simultaneously queried. Additionally, we define “useful” as responding within 2x the time it takes any other peer from our routing table to respond to us (this biases against peers that are slow, overloaded, unreliable, or have bad network connectivity to us). The definitions of replaceable and useful peers are likely to change as we gather more information about the dynamics of the network and investigate relevant threat models.
```

File: content/post/104-dht-deep-dive.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    *   Used by IPNS
+*   Peer Records: These map a peerID to a set of multiaddresses at which the peer may be reached
+    *   Used by IPFS when we know of a peer with content, but do not know its address
+    *   Used for manual connections (e.g., `ipfs swarm connect /p2p/QmXYZ`)
+
+Each of these record types has slightly different semantics, but they are all updated and found via the same DHT protocol, IPFS’s take on Kademlia.
+
+## Kademlia Overview
+
+The Kademlia algorithm has been around for a while and there are some great resources already available for it online, including the paper itself [link Kademlia paper on IPFS] and Wikipedia [link Kademlia wikipedia]. We’ll go through some of the basics here, though.
```
[the remark this answers is not in the document]
Comment (aschmahmann) [reply content/post/104-dht-deep-dive.md:35]: should the link above be ipfs.io or dweb.link?

[the remark this answers is not in the document]
Comment (momack2) [reply content/post/104-dht-deep-dive.md:35]: this works =]

File: content/post/104-dht-deep-dive.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+To make things easier for this use case, we have _added a second DHT_ that is meant to include nodes that are not part of the public network, such as VPNs, CJDNS, Yggdrasil, etc. For now, we’ll refer to this as the LAN DHT, as opposed to the public network which is the WAN DHT. These two DHTs are separated by utilizing different DHT protocol names (i.e., `/ipfs/kad/1.0.0` for the WAN DHT, and `/ipfs/lan/kad/1.0.0` for the LAN DHT) to eliminate any accidental merging of the two networks. However, all of the non-public networks do have some risk of merging if users have not properly configured their networks.
+
+The main implementation difference between the WAN and LAN DHTs are the acceptance criteria for peers—which are eligible to be part of a routing table or query and which are not. The WAN DHT’s criteria is “do you look like a public address” and the LAN DHT’s criteria is “do you look like a non-public address.” However, while WAN DHT nodes switch from client to server mode based on whether they are publicly dialable, nodes are always LAN DHT servers (unless the “dhtclient” option has been set).
```
Comment (petar) [inline content/post/104-dht-deep-dive.md:69]: Sentence symmetry, perhaps:

However, while WAN DHT nodes switch from client to server mode based on whether they are publicly dialable, nodes are always LAN DHT servers (unless the “dhtclient” option has been set).
->
However, while WAN DHT nodes switch from client to server mode based on whether they are publicly dialable, LAN DHT nodes are always servers (unless the “dhtclient” option has been set).

File: content/post/104-dht-deep-dive.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+2. We also search for ourselves in the network (i.e., bucket 255), just in case the network size and distribution are such that the first 15 buckets do not suffice for us to learn about the `K` peers closest to us.
+
+## Lookup Algorithm
+
+The lookup algorithm answers the question “What are the `K` closest peers to `X`?”. Our implementation of the Kademlia lookup algorithm goes like this:
+
+1. Load the `K` closest peers to `X` from our routing table into the query queue
+2. Allowing up to Alpha concurrent queries (Alpha in go-ipfs is 10, but is an implementation parameter not inherent in the network itself) grab the peer closest to `X` and ask them “who are the `K` closest peers to `X`?”
+3. When a query to a peer completes add those results to the query queue, and pull the next peer off the queue and query them
+4. The query terminates whenever the closest Beta peers to `X` have been successfully queried (i.e. no dial timeouts, errors, etc.).
```
Comment (petar) [inline content/post/104-dht-deep-dive.md:99]: closest -> closest known

File: content/post/104-dht-deep-dive.md (hunk 4 above)
Comment (petar) [reply content/post/104-dht-deep-dive.md]: This looks fine. It introduces "replaceable" to mean "not recently useful". And the definition of "recently" and "useful" is unchanged.

File: content/post/104-dht-deep-dive.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+Note that the selection of network wide parameters like `K` is not arbitrary, but is instead determined based on the observed/expected churn in the network and the frequency with which the network will republish information. These system and network parameters drive the decisions made in Kademlia’s two main components, the routing table which tracks all those links in the network, and the lookup algorithm which determines how to traverse those links to store and retrieve data.
```
[the remark this answers is not in the document]
Comment (aschmahmann) [reply content/post/104-dht-deep-dive.md:45]: > System parameters (like K) are computed to maximize the probability that the network stays connected and no data is lost, while maintaining a desired latency for queries, and assuming the observations (of average churn) stay constant.

This seems reasonable do you mind adding it where you'd like as a suggestion in GitHub (the +- button in the text box) so that it's easy to batch the updates?

File: content/post/104-dht-deep-dive.md (hunk 7 above)
Comment (aschmahmann) [reply content/post/104-dht-deep-dive.md:99]:
```suggestion
4. The query terminates whenever the closest known Beta peers to `X` have been successfully queried (i.e. no dial timeouts, errors, etc.).
```

File: content/post/104-dht-deep-dive.md (hunk 6 above)
Comment (aschmahmann) [reply content/post/104-dht-deep-dive.md:69]:
```suggestion
The main implementation difference between the WAN and LAN DHTs are the acceptance criteria for peers—which are eligible to be part of a routing table or query and which are not. The WAN DHT’s criteria is “do you look like a public address” and the LAN DHT’s criteria is “do you look like a non-public address.” However, while WAN DHT nodes switch from client to server mode based on whether they are publicly dialable, LAN DHT nodes are always servers (unless the “dhtclient” option has been set).
```

File: content/post/104-dht-deep-dive.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+*   A projection, that will take a record key and calculate a position in the address space where the peer(s) most ideally suited to store the record should be near (in IPFS we use `SHA256(Record Key)`)
+
+Having this address space and a peer ordering metric allows us to search the network as though it was a sorted list. In particular, we can turn the system into something like a skip list where a peer knows peers that are distances of around 1,2,4,8,16… away from it. This will allow us to search the list in time that is logarithmic in the size of the network (i.e. O(log(N)) lookup time). Unlike a skip list, Kademlia is somewhat unstable since peers can join, leave and rejoin the network at any time. To deal with the unstable nature of the system, a Kademlia peer does not just keep links to the peers with distance 1,2,4,8,... away from it, but instead for each multiple of 2 away keeps up to `K` (in IPFS 20) links. For example, instead of a peer keeping a single link 128 away it would keep 20 links that are between 65 and 128 away.
```
[the remark this answers is not in the document]
Comment (aschmahmann) [reply content/post/104-dht-deep-dive.md:43]:
```suggestion
Having this address space and a peer ordering metric allows us to search the network as though it was a sorted list. In particular, we can turn the system into something like a skip list where a peer knows peers that are distances of around 1,2,4,8,16… away from it. This will allow us to search the list in time that is logarithmic in the size of the network (i.e. O(log(N)) lookup time). Unlike a skip list, Kademlia is somewhat unstable since peers can join, leave and rejoin the network at any time. To deal with the unstable nature of the system, a Kademlia peer does not just keep links to the peers with distance 1,2,4,8,... away from it, but instead for each multiple of 2 away it keeps up to `K` (in IPFS 20) links. For example, instead of a peer keeping a single link 128 away it would keep 20 links that are between 65 and 128 away.
```

File: content/post/104-dht-deep-dive.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (petar) [reply content/post/104-dht-deep-dive.md:45]:
```suggestion
Note that the selection of network wide parameters like `K` is not arbitrary, but is instead determined based on the observed/expected churn in the network and the frequency with which the network will republish information. System parameters (like K) are computed to maximize the probability that the network stays connected and no data is lost, while maintaining a desired latency for queries, and assuming the observations (of average churn) stay constant. These system and network parameters drive the decisions made in Kademlia’s two main components, the routing table which tracks all those links in the network, and the lookup algorithm which determines how to traverse those links to store and retrieve data.
```

File: content/post/104-dht-deep-dive.md [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+Note that the selection of network wide parameters like `K` is not arbitrary, but is instead determined based on the observed/expected churn in the network and the frequency with which the network will republish information. These system and network parameters drive the decisions made in Kademlia’s two main components, the routing table which tracks all those links in the network, and the lookup algorithm which determines how to traverse those links to store and retrieve data.
+
+## Kademlia and Undialable Peers
+
+As noted above, a major property of Kademlia is that all peers can be assembled on to a line from smallest to largest. This property is useful since it means that as peer 0 walks down the line to find peer 55 it can know it is getting progressively closer. However, this requires that everyone on the line can talk to each other, otherwise peer 33 might send peer 0 down a dead end by telling them the content they want is on a node they can’t communicate with. This can result in the network being slow, and more importantly, **fragmented** with data being accessible by some peers and not others.
```
Comment (jacobheun) [inline content/post/104-dht-deep-dive.md:49]: "on to a line" -> "inline"
```suggestion
As noted above, a major property of Kademlia is that all peers can be assembled inline from smallest to largest. This property is useful since it means that as peer 0 walks down the line to find peer 55 it can know it is getting progressively closer. However, this requires that everyone on the line can talk to each other, otherwise peer 33 might send peer 0 down a dead end by telling them the content they want is on a node they can’t communicate with. This can result in the network being slow, and more importantly, **fragmented** with data being accessible by some peers and not others.
```

File: content/post/104-dht-deep-dive.md [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+To do this, we use libp2p’s AutoNAT, which acts as a distributed STUN layer informing peers of their observed addresses and whether or not they appear to be publicly dialable. Only when peers detect that they are publicly dialable do they switch from client mode (where they can query the DHT, but not respond to queries) to server mode (where they can both query and respond to queries). Similarly, if a server discovers that it is no longer publicly dialable it will switch back into client mode.
+
+Serving AutoNAT requests (i.e., checking whether other peers were dialable) was previously only enabled on opt-in nodes like some of the IPFS public infrastructure. However, leaning so heavily on AutoNAT to clean up the DHT from undialable nodes made us really want to make AutoNAT more distributed. Therefore, we now expose a rate-limited AutoNAT service on all IPFS nodes that have discovered that they are publicly dialable. These requests should be quite infrequent and therefore not have noticeable overhead for standard IPFS nodes.
```
Comment (jacobheun) [inline content/post/104-dht-deep-dive.md:57]: Some readability cleanup.

```suggestion
Serving AutoNAT requests (i.e., checking whether other peers were dialable) was previously only enabled on opt-in nodes like some of the IPFS public infrastructure. However, leaning so heavily on AutoNAT to clean up undialable nodes from the DHT made us push to make AutoNAT more distributed. Therefore, we now expose a rate-limited AutoNAT service on all IPFS nodes that have discovered that they are publicly dialable. These requests should be infrequent and therefore not have noticeable overhead for standard IPFS nodes.
```

File: content/post/104-dht-deep-dive.md [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+2. We also search for ourselves in the network (i.e., bucket 255), just in case the network size and distribution are such that the first 15 buckets do not suffice for us to learn about the `K` peers closest to us.
+
+## Lookup Algorithm
+
+The lookup algorithm answers the question “What are the `K` closest peers to `X`?”. Our implementation of the Kademlia lookup algorithm goes like this:
+
+1. Load the `K` closest peers to `X` from our routing table into the query queue
+2. Allowing up to Alpha concurrent queries (Alpha in go-ipfs is 10, but is an implementation parameter not inherent in the network itself) grab the peer closest to `X` and ask them “who are the `K` closest peers to `X`?”
+3. When a query to a peer completes add those results to the query queue, and pull the next peer off the queue and query them
```
Comment (jacobheun) [inline content/post/104-dht-deep-dive.md:98]: Clarify the queue is sorted by closeness.
```suggestion
3. When a query to a peer completes add those results to the query queue, and pull the next, closest peer off the queue and query them
```

File: content/post/104-dht-deep-dive.md [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+As can be seen in the graphs, the most drastic changes are to 95th percentile lookup times and to the operations that spent more time doing their lookups and could not terminate as early. This meant IPFS Provide and IPNS Put which require actually completing a search through the network got a very large boost (for Provide 24x speedup on average and 33x speedup for the 95th percentile). This was followed by IPNS Get which needs to find many records, then Find Peer which is looking for one very specific record, and finally the time to find just one IPFS Provider record was sped up by 2.2x on average and 6.4x for the 95th percentile.
```
Comment (jacobheun) [inline content/post/104-dht-deep-dive.md:135]:
```suggestion
As can be seen in the graphs, the most drastic changes are to 95th percentile lookup times and to the operations that spent more time doing their lookups and could not terminate as early. This meant IPFS Provide and IPNS Put, which require actually completing a search through the network, got a very large boost (for Provide 24x speedup on average and 33x speedup for the 95th percentile). This was followed by IPNS Get which needs to find many records, then Find Peer which is looking for one very specific record, and finally the time to find just one IPFS Provider record was sped up by 2.2x on average and 6.4x for the 95th percentile. 
```

File: content/post/104-dht-deep-dive.md [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+
+Throughout the development process we ran many Testground tests to get an understanding of how our changes have improved the network. Below is a comparison of the performance of a 1000 peer network where all peers have around 100-120ms latencies from each other, that is running the DHT from go-ipfs v0.4.23 and the DHT from go-ipfs v0.5.0. A few notes: The v0.4.23 DHT had a few modifications to make testing easier like removing hard coded lookup timeouts, so we see just how long the queries should really be running for. Additionally, while the v0.4.23 test had 60% of the peers be undialable, the v0.5.0 test had 0 undialable nodes. This change actually benefits the v0.4.23 results, since in v0.5.0 if 60% of the nodes are undialable then, because v0.5.0 undialable nodes no longer advertise themselves as DHT servers, the network itself effectively shrinks to 40% of its size, which decreases lookup times.
```
Comment (jacobheun) [inline content/post/104-dht-deep-dive.md:133]: > Additionally, while the v0.4.23 test had 60% of the peers be undialable, the v0.5.0 test had 0 undialable nodes. This change actually benefits the v0.4.23 results, since in v0.5.0 if 60% of the nodes are undialable then, because v0.5.0 undialable nodes no longer advertise themselves as DHT servers, the network itself effectively shrinks to 40% of its size, which decreases lookup times.

This is not clear and is a bit hard to follow. It states 0.4.23 had an advantage in the test suite, but the followup statements make it sound like 0.5 had the benefit in testing.

If this isn't easy to clarify succinctly I would just remove it.

File: content/post/104-dht-deep-dive.md [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+author: Adin Schmahmann
+header_image: 104-dht-deep-dive.png
+tags: go-ipfs, 0.5 release
+---
+
+<p style="max-width:1000px;margin-left:auto;margin-right:auto;">
+  <img src="/header_images/104-dht-deep-dive.png">
+</p>
+
+At the end of April, we released our largest update to go-ipfs to date: [IPFS 0.5](https://blog.ipfs.io/2020-04-28-go-ipfs-0-5-0/). While there have been many improvements, the changes to IPFS’s Distributed Hash Table (DHT) were especially critical to improving the performance and stability of finding data in IPFS. For some background on how we arrived at the most recent set of DHT changes, take a look at [The Road to the New DHT](https://blog.ipfs.io/2020-05-19-road-to-dht/).
```
Comment (jacobheun) [inline content/post/104-dht-deep-dive.md:14]: I know we mention upgrading at the end, but I think the intro is also a good place to plug upgrading.

```suggestion
At the end of April, we released our largest update to go-ipfs to date: [IPFS 0.5](https://blog.ipfs.io/2020-04-28-go-ipfs-0-5-0/). While there have been many improvements, the changes to IPFS’s Distributed Hash Table (DHT) were especially critical to improving the performance and stability of finding data in IPFS. For some background on how we arrived at the most recent set of DHT changes, take a look at [The Road to the New DHT](https://blog.ipfs.io/2020-05-19-road-to-dht/), or try them out yourself in the [latest version of go-ipfs](https://dist.ipfs.io/#go-ipfs).
```

File: content/post/104-dht-deep-dive.md (hunk 14 above)
Comment (aschmahmann) [reply content/post/104-dht-deep-dive.md:133]: I'm fine removing it, it was more of a "full disclosure" kind of thing.

The v0.4.23 test had undialable nodes in it and the v0.5.0 did not. That might seem like an advantage to the v0.5.0 test since undialable=bad, but because of our work on v0.5.0 we actually biased the results towards helping v0.4.23 and still trounced it easily.

Comment (jacobheun) [reply content/post/104-dht-deep-dive.md:133]: I think removing it is fine since it already mentions needing to tweak 0.4.23 for testing purposes.

Comment (aschmahmann) [reply content/post/104-dht-deep-dive.md:133]:
```suggestion
Throughout the development process we ran many Testground tests to get an understanding of how our changes have improved the network. Below is a comparison of the performance of a 1000 peer network where all peers have around 100-120ms latencies from each other, that is running the DHT from go-ipfs v0.4.23 and the DHT from go-ipfs v0.5.0. A few notes: The v0.4.23 DHT had a few modifications to make testing easier like removing hard coded lookup timeouts, so we see just how long the queries should really be running for.
```

File: content/post/104-dht-deep-dive.md [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+        *   We will try to connect to the peer with ID `H` as soon as we learn addresses about it. The lookup can terminate early if we end up connecting to the peer.
+
+## Testing and Results
+
+There have been a lot of changes to the DHT as part of the go-ipfs v0.5.0 release. While many of the changes were intuitively going to be quite useful, we needed harder evidence that the full set of changes would result in a stable and performant network. To that end, we utilized Testground, which is a new distributed testing infrastructure (check out their launch notes at [Testground blog post](https://blog.ipfs.io/2020-05-06-launching-testground/)). 
+
+Throughout the development process we ran many Testground tests to get an understanding of how our changes have improved the network. Below is a comparison of the performance of a 1000 peer network where all peers have around 100-120ms latencies from each other, that is running the DHT from go-ipfs v0.4.23 and the DHT from go-ipfs v0.5.0. A few notes: The v0.4.23 DHT had a few modifications to make testing easier like removing hard coded lookup timeouts, so we see just how long the queries should really be running for.
```
Comment (aschmahmann) [inline content/post/104-dht-deep-dive.md:133]:
```suggestion
Throughout the development process we ran many Testground tests to get an understanding of how our changes have improved the network. Below is a comparison of the performance of a 1000 peer network where all peers have around 100-120ms latencies from each other, that is running the DHT from go-ipfs v0.4.23 and the DHT from go-ipfs v0.5.0. Note: The v0.4.23 DHT had small modifications to make testing easier like removing hard coded lookup timeouts, so we can see just how long the queries should really be running for.
```
